src/test/regress/sql/multi_remove_node_reference_table.sql

```diff
@@ -0,0 +1,386 @@
+--
+-- MULTI_REMOVE_NODE_REFERENCE_TABLE
+--
+-- Tests that check the metadata after master_remove_node.
+
+
+ALTER SEQUENCE pg_catalog.pg_dist_shardid_seq RESTART 1380000;
+ALTER SEQUENCE pg_catalog.pg_dist_jobid_seq RESTART 1380000;
+ALTER SEQUENCE pg_catalog.pg_dist_colocationid_seq RESTART 1380000;
+
+-- create copy of pg_dist_shard_placement to reload after the test
+CREATE TABLE tmp_shard_placement AS SELECT * FROM pg_dist_shard_placement WHERE nodeport = :worker_2_port;
+DELETE FROM pg_dist_shard_placement WHERE nodeport = :worker_2_port;
+
+
+-- remove non-existing node
+SELECT master_remove_node('localhost', 5555);
+
+
+-- remove a node with no reference tables
+
+-- verify node exist before removal
+SELECT COUNT(*) FROM pg_dist_node WHERE nodeport = :worker_2_port;
+
+SELECT master_remove_node('localhost', :worker_2_port);
+
+-- verify node is removed
+SELECT COUNT(*) FROM pg_dist_node WHERE nodeport = :worker_2_port;
+
+-- re-add the node for next tests
+SELECT master_add_node('localhost', :worker_2_port);
+
+
+-- remove a node with reference table with active placement
+CREATE TABLE remove_node_reference_table(column1 int);
+SELECT create_reference_table('remove_node_reference_table');
+SELECT master_remove_node('localhost', :worker_2_port);
+
+
+-- remove a node with reference table with no active placement
+UPDATE pg_dist_shard_placement SET shardstate = 3 WHERE nodeport = :worker_2_port;
+
+-- status before master_remove_node
+SELECT COUNT(*) FROM pg_dist_node WHERE nodeport = :worker_2_port;
+
+SELECT
+    shardid, shardstate, shardlength, nodename, nodeport
+FROM
+    pg_dist_shard_placement
+WHERE
+    nodeport = :worker_2_port;
+
+SELECT *
+FROM pg_dist_colocation
+WHERE colocationid IN
+    (SELECT colocationid
+     FROM pg_dist_partition
+     WHERE logicalrelid = 'remove_node_reference_table'::regclass);
+
+SELECT master_remove_node('localhost', :worker_2_port);
+
+-- status after master_remove_node
+SELECT COUNT(*) FROM pg_dist_node WHERE nodeport = :worker_2_port;
+
+SELECT
+    shardid, shardstate, shardlength, nodename, nodeport
+FROM
+    pg_dist_shard_placement
+WHERE
+    nodeport = :worker_2_port;
+
+SELECT *
+FROM pg_dist_colocation
+WHERE colocationid IN
+    (SELECT colocationid
+     FROM pg_dist_partition
+     WHERE logicalrelid = 'remove_node_reference_table'::regclass);
+
+
+-- remove same node twice
+SELECT master_remove_node('localhost', :worker_2_port);
+
+-- re-add the node for next tests
+SELECT master_add_node('localhost', :worker_2_port);
+UPDATE pg_dist_shard_placement SET shardstate = 3 WHERE nodeport = :worker_2_port;
```

Comment:
As I noted earlier, we should not need to set `shardstate` to `3`. In fact, it is never possible to set shard state of reference tables to `3` for reference tables.